### PR TITLE
add --classweight

### PR DESCRIPTION
**Pull request by denik**
$ vw --classweight LABEL1:WEIGHT1[,LABEL2:WEIGHT2...]

Multiplies the weight of the examples with label equal to LABEL1 by WEIGHT1, examples with label equal to LABEL2 by WEIGHT2 and so on.

This is already possible to achieve by rewriting the input file, however, making this built-in makes it much faster to experiment with different class weights.

Note, that this does affect the 'average loss' reported, the same way as modifying the examples' weight would.

### CODE REVIEW

**Comment by JohnLangford** [comment]
This seems like a good feature idea but the implementation is antimodular.  Could I ask you to refactor things so that classweights is a reduction instead of a modification to the parser?  Then you could stick the reduction about here: https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/parse_args.cc#L1128

**Comment by denik** [comment]
I've moved all the code to a reduction. Can you take a look?

Note, that I had to replace usage of ec.l.multi.weight in oaa.cc and recall_tree.cc with ec.weight. It seems that ec.weight should be always the same, but maybe I'm wrong about it.

File: vowpalwabbit/label_parser.h [hunk 1]
```diff
@@ -16,4 +16,5 @@ struct label_parser
   float (*get_weight)(void*);
   void (*copy_label)(void*,void*); // copy_label(dst,src) performs a DEEP copy of src into dst (dst is allocated correctly).  if this function is nullptr, then we assume that a memcpy of size label_size is sufficient, so you need only specify this function if your label constains, for instance, pointers (otherwise you'll get double-free errors)
   size_t label_size;
+  int (*get_int_label)(void*);
```
Comment (JohnLangford) [inline vowpalwabbit/label_parser.h]: Why is this needed?

File: vowpalwabbit/oaa.cc [hunk 2]
```diff
@@ -133,7 +135,7 @@ void finish_example_scores(vw& all, oaa& o, example& ec)
   { if (ec.l.multi.label <= o.k) // prevent segmentation fault if labeĺ==(uint32_t)-1
       correct_class_prob = ec.pred.scalars[ec.l.multi.label-1];
     if (correct_class_prob > 0)
-      multiclass_log_loss = -log(correct_class_prob) * ec.l.multi.weight;
+      multiclass_log_loss = -log(correct_class_prob) * ec.weight;
```
Comment (JohnLangford) [inline vowpalwabbit/oaa.cc]: This is fixing a bug.

File: vowpalwabbit/label_parser.h (hunk 1 above)
Comment (denik) [reply vowpalwabbit/label_parser.h]: In order to apply class weight, we need to know example's label. This is a way to get the label regardless of example type (simple_label and multiclass store label differently).

File: vowpalwabbit/multiclass.cc [hunk 3]
```diff
@@ -86,7 +91,8 @@ label_parser mc_label = {default_label, parse_label,
                          cache_label, read_cached_label,
                          delete_label, weight,
                          nullptr,
-                         sizeof(label_t)
+                         sizeof(label_t),
+                         get_int_label,
```
Comment (JohnLangford) [inline vowpalwabbit/multiclass.cc]: Removable?

File: vowpalwabbit/parse_args.cc [hunk 4]
```diff
@@ -1144,6 +1145,7 @@ void parse_reductions(vw& all)
   all.reduction_stack.push_back(bs_setup);
 
   all.reduction_stack.push_back(audit_regressor_setup);
+  all.reduction_stack.push_back(classweight_setup);
```
Comment (JohnLangford) [inline vowpalwabbit/parse_args.cc]: This is to late in the reduction stack.   Can you shift to the spot I pointed out?

File: vowpalwabbit/simple_label.cc [hunk 5]
```diff
@@ -38,6 +38,11 @@ float get_weight(void* v)
   return ld->weight;
 }
 
+static int get_int_label(void* v)
```
Comment (JohnLangford) [inline vowpalwabbit/simple_label.cc]: Removable?

File: vowpalwabbit/multiclass.cc (hunk 3 above)
Comment (denik) [reply vowpalwabbit/multiclass.cc]: Not sure how to get example label otherwise for different label types. They all stored in a union, that makes it a bit complicated.

File: vowpalwabbit/oaa.cc (hunk 2 above)
Comment (denik) [reply vowpalwabbit/oaa.cc]: should I commit it separately?

**Comment by JohnLangford** [comment]
This seems much closer, but it looks like there is still some abstraction leakage.  Removable?

**Comment by JohnLangford** [comment]
The appveyor build can presumably be fixed by tweaking the windows build for the new file just above and below here: https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/vw_dynamic.vcxproj#L436

File: vowpalwabbit/label_parser.h (hunk 1 above)
Comment (JohnLangford) [reply vowpalwabbit/label_parser.h]: The traditional method would be get the datatype in the arguments, but we actually have some type information here:
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/learner.h#L113 
now.  

You should be able to create the base learner and then observe it's type.  It's also important that the reduction set it's type appropriately with the init() call.

File: vowpalwabbit/oaa.cc (hunk 2 above)
Comment (JohnLangford) [reply vowpalwabbit/oaa.cc]: No, it's fine.

**Comment by denik** [comment]
I've pushed an update, please take a look.

File: vowpalwabbit/label_parser.h (hunk 1 above)
Comment (denik) [reply vowpalwabbit/label_parser.h]: ah, do you mean add a new templater parameter to predict_or_learn that specifies how to get the label? so far I've done it at runtime. I can update to templated one.

**Comment by JohnLangford** [comment]
The code looks good now, thanks :-)  I'll merge after automated tests succeed.

File: vowpalwabbit/label_parser.h (hunk 1 above)
Comment (JohnLangford) [reply vowpalwabbit/label_parser.h]: I'd template myself---it should never vary once the reduction stack is instantiated.

**Comment by denik** [comment]
I've templated the switch on pred_type in https://github.com/JohnLangford/vowpal_wabbit/pull/1348/commits/28eb92c92f27141cb7209808b94c1fcd1ee8bec6

**Comment by JohnLangford** [comment]
Merged, thank you.
